helix-vcs/src/rope.rs

```diff
@@ -0,0 +1,20 @@
+use std::hash::Hash;
+
+use ropey::Rope;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct RopeLine<'a>(pub ropey::RopeSlice<'a>);
+
+impl Hash for RopeLine<'_> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        for chunk in self.0.chunks() {
+            chunk.hash(state);
+        }
+    }
+}
```

Comment:
I've now implemented `Hash` directly on `Rope` and `RopeSlice` (in the recently released Ropey 1.4.0), so this should be unnecessary now.

It's also worth noting that your implementation here is incorrect (not to criticize, just an interesting technical note), which I only noticed when diving into the trait to implement it myself:

`&str`'s `Hash` implementation, which you're calling in the loop here, is purposefully implemented so that hashing a series of string slices yields a different hash than hashing the same data as a single concatenated string.  Which means that if two `Rope`s have identical contents, but the chunks happen to be split up differently, your implementation will give two different hashes for them.

Instead, you actually need to call `state.write(chunk.as_bytes())` in the inner loop.

The reason `&str`'s `Hash` implementation works like that is because of `#[derive(Hash)]`.  Imagine this struct:

```rust
#[derive(Hash)]
struct Person<'a> {
    first_name: &'a str,
    second_name: &'a str,
}
```

If concatenated strings gave the same hash, then someone named "Jon Athanbor" and "Jonathan Bor" would has to the same value if the names were stored in lower case.  (A contrived example, I know, but it's just to illustrate.)

Reply:
Yeah usually you prefix each chunk with a hash of the `len` to solve for this

Reply:
Ah, I didn't know this; thanks for pointing it out ! Thanks for the Hash impl too :)